keps/sig-storage/20191031-storage-capacity-constraints-for-pod-scheduling.md

```diff
@@ -0,0 +1,894 @@
+  for one CSI driver, so overall size in etcd is higher.
+- Higher number of objects which all need to be retrieved by a client
+  which does not already know which `CSIStoragePool` has a certain spec.
+- Constraints like "unique spec" cannot be enforced by the API server.
+
+##### Example: local storage
+
+```
+apiVersion: storage.k8s.io/v1alpha1
+kind: CSIStoragePool
+metadata:
+  name: sp-ab96d356-0d31-11ea-ade1-8b7e883d1af1
+spec:
+  driverName: hostpath.csi.k8s.io
+  nodes:
+  - node-1
+status:
+  classes:
+  - capacity: 256G
+    storageClassName: <fallback>
+
+apiVersion: storage.k8s.io/v1alpha1
+kind: CSIStoragePool
+metadata:
+  name: sp-c3723f32-0d32-11ea-a14f-fbaf155dff50
+spec:
+  driverName: hostpath.csi.k8s.io
+  nodes:
+  - node-2
+status:
+  classes:
+  - capacity: 512G
+    storageClassName: <fallback>
+```
+
+##### Example: affect of storage classes
+
+```
+apiVersion: storage.k8s.io/v1alpha1
+kind: CSIStoragePool
+metadata:
+  name: sp-9c17f6fc-6ada-488f-9d44-c5d63ecdf7a9
+spec:
+  driverName: lvm
+  nodes:
+  - node-1
+status:
+  classes:
+  - capacity: 256G
+    storageClassName: striped
+  - capacity: 128G
+    storageClassName: mirrored
+```
+
+##### Example: network attached storage
+
+```
+apiVersion: storage.k8s.io/v1alpha1
+kind: CSIStoragePool
+metadata:
+  name: sp-b0963bb5-37cf-415d-9fb1-667499172320
+spec:
+  driverName: pd.csi.storage.gke.io
+  nodeTopology:
+    nodeSelectorTerms:
+    - matchExpressions:
+      - key: failure-domain.beta.kubernetes.io/region
+        operator: In
+        values:
+        - us-east-1
+status:
+  classes:
+  - capacity: 128G
+    storageClassName: <fallback>
+
+apiVersion: storage.k8s.io/v1alpha1
+kind: CSIStoragePool
+metadata:
+  name: sp-64103396-0d32-11ea-945c-e3ede5f0f3ae
+spec:
+  driverName: pd.csi.storage.gke.io
+status:
+  classes:
+  - capacity: 128G
+    storageClassName: <fallback>
+  nodeTopology:
+    nodeSelectorTerms:
+    - matchExpressions:
+      - key: failure-domain.beta.kubernetes.io/region
+        operator: In
+        values:
+        - us-west-1
+status:
+  classes:
+  - capacity: 256G
+    storageClassName: <fallback>
+```
+
+#### CSIDriver.spec.storageCapacity
+
+A new field `storageCapacity` of type `boolean` with default `false`
+in
+[CSIDriver.spec](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.16/#csidriverspec-v1beta1-storage-k8s-io)
+indicates whether a driver deployment will create `CSIStoragePool`
+objects. This allows the scheduler to determine whether it needs to
+consider available capacity for volumes of that driver or can proceed
+without it (as it does now).
+
+
+### Updating capacity information with external-provisioner
+
+Most (if not all) CSI drivers already get deployed on Kubernetes
+together with the external-provisioner which then handles volume
+provisioning via PVC. However, drivers which only handle inline
+ephemeral volumes currently don’t need that and thus get deployed
+without it.
+
```

Comment:
Although the core concept of extending external-provisioner is simple, to consider ephemeral volume, this KEP seems to complicate external-volume very much.

First, you introduce a new option `--enable-capacity` for two different concepts, the kind of volume and topology-awareness. I think topology-awareness can be distinguished by VOLUME_ACCESSIBILITY_CONSTRAINTS capability. In addition, the option can be `--enable-ephemeral-volume={true,false} (default is false).`
Second, since the use case of "only local volume" is very specific, defining the spec only for it looks overkill.
Last, it doesn't consider the non-local volume and the non-topology-aware case.

So, in my understanding, adding the following spec is sufficient.

- Introduce `--enable-capacity` option as described above.
- Define the behavior of both non-topology-aware case and topology-aware case.

It can simplify this KEP very much.

Reply:
> I think topology-awareness can be distinguished by VOLUME_ACCESSIBILITY_CONSTRAINTS capability. 

That doesn't tell external-provisioner what exactly the accessibility constraints are. If I understand you correctly, you are proposing to remove the special case for node-local storage and just treat all drivers with topology support the same way, right?

That would work, but I think the special case with node-local storage has some potential for optimizing the execution (no need to watch nodes and calculate topology in the external-provisioner, simpler lookup of suitable nodes in the scheduler) and the result (shorter `CSIStoragePool` instances) that it is worthwhile to support it.

I'm open for suggestions for naming the command line options. Instead of `--enable-capacity=local-ephemeral` we could also add `--enable-capacity-local-ephemeral` (single bool). This is a bit repetitive, though, which is why I picked the more complicated `--enable-capacity`.

> Last, it doesn't consider the non-local volume and the non-topology-aware case.

Indeed, that is an oversight. That case can be supported with `CSIStoragePool` instances that has neither `Nodes` nor `NodeTopology` set ("If neither is set, the pool is assumed
to be available in the entire cluster."), it's just the external-provisioner which needs to know what to create.

Reply:
> If I understand you correctly, you are proposing to remove the special case for node-local storage and just treat all drivers with topology support the same way, right?

Yes.

BTW, when I reconsider this idea, I came up with a fundamental question. In my understanding, node-exported is to help to triggers Controller{Create,Delete}Volume. However, ephemeral volume uses only Node{Publish,Unpublish}.

Could you tell me the relationship between ephemeral inline volume and the external-provisioner that is introduced by this KEP?

Reply:
For ephemeral volumes, none of the usual functionality of the external-provisioner (provisioning aka calling CreateVolume and its inverse) are needed. In the context of this KEP, external-provisioner is just a convenient place to put the necessary new functionality.

We could instead introduce a new "external-capacity-tracker". I just don't think that yet another sidecar is the right way to go. Releasing the current set is already causing enough work and complications. I'd rather add some new optional features to an existing one. I picked the external-provisioner because it is the one that most CSI drivers will have; even the deployment where it runs alongside each node service was already something that came up earlier.

Reply:
I removed it again based on @msau42's feedback that it would make the external-provisioner too complex, see 86cf87dc7a2e58fcfec9034180d14ef6bab29fcb.